util/whitelist/debugger_whitelist.go

```diff
@@ -0,0 +1,83 @@
+package whitelist
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/patrickmn/go-cache"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	whiteListURL   string
+	whiteListCache = cache.New(5*time.Minute, 5*time.Minute)
+)
+
+func init() {
+	whiteListURL = os.Getenv("WHITELIST_BACKEND_URL")
+	logrus.Info("whiteListURL: ", whiteListURL)
+}
+
+// IsIPValid checks if the debugger IP is in the whitelist through the whitelist backend
+func IsIPValid(ip string) (bool, error) {
+	if ip == "unknown_ip" {
+		return false, nil
+	}
+
+	cacheKey := "whitelist-ip-" + ip
+	cacheValue, found := whiteListCache.Get(cacheKey)
+	logrus.Debug("whitelist IP cache Get ip: ", ip, ", found: ", found, ", cacheValue: ", cacheValue)
+	if found {
+		return cacheValue.(bool), nil
+	}
+
+	params := url.Values{}
+	Url, err := url.Parse(whiteListURL + "api/get_debugger")
+	if err != nil {
+		return false, err
+	}
+	params.Set("accept", "application/json")
+	params.Set("ip", ip)
+	Url.RawQuery = params.Encode()
+	urlPath := Url.String()
+	resp, err := http.Get(urlPath)
+	if err != nil {
+		return false, err
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return false, err
+	}
+	var data map[string]interface{}
+	err = json.Unmarshal(body, &data)
+	if err != nil {
+		return false, err
+	}
+
+	debuggerList, ok := data["debugger"]
+	isValid := (ok == true) && (debuggerList != nil)
+	whiteListCache.Set(cacheKey, isValid, cache.DefaultExpiration)
+	return isValid, nil
+}
+
+func GetIPFromRequestEnv(req *http.Request) string {
+	fwdAddress := req.Header.Get("X-Forwarded-For")
```

Comment:
Can these headers be spoofed by a malicious actor?

cc @iczc

Reply:
> Can these headers be spoofed by a malicious actor?
> 
> cc @iczc

yes, there is an ip forgery issue here.

1. If the rpc gateway is not preceded by a reverse proxy, the IP  should get from `ip = strings.Split(req.RemoteAddr, ":")[0]`,  but a malicious actor can set the X-Forwarded-For in the request header to forge IP.
`X-Forwarded-For: illegalIp`

2. This issue cannot be solved if the rpc gateway uses reverse proxies. Each reverse proxy appends the request IP to the X-Forwarded-For header in order, and the IP forged by the malicious in the X-Forwarded-For header will be treated by the first reverse proxy as the previous reverse proxy‘s result. In this case, `strings.Split(fwdAddress, ", ")[0]` is the malicious' forged IP. 
`X-Forwarded-For: illegalIp, client, proxy1..proxyN`

The recommended implementation can be found at
https://github.com/scroll-tech/eth-faucet/blob/v1.0.0/internal/server/limiter.go#L77-L102 , 
solved by configuring the count of reverse proxy:

```golang
func getClientIPFromRequest(proxyCount int, r *http.Request) string {
	if proxyCount > 0 {
		xForwardedFor := r.Header.Get("X-Forwarded-For")
		xRealIP := r.Header.Get("X-Real-Ip")

		if xForwardedFor != "" {
			xForwardedForParts := strings.Split(xForwardedFor, ",")
			// Avoid reading the user's forged request header by configuring the count of reverse proxies
			partIndex := len(xForwardedForParts) - proxyCount
			if partIndex < 0 {
				partIndex = 0
			}
			return strings.TrimSpace(xForwardedForParts[partIndex])
		}

		if xRealIP != "" {
			return strings.TrimSpace(xRealIP)
		}
	}

	remoteIP, _, err := net.SplitHostPort(r.RemoteAddr)
	if err != nil {
		remoteIP = r.RemoteAddr
	}
	return remoteIP
}
```

Reply:
Yes and thanks! Noticed this when glanced eth-faucet repo, will change it to this implementation.

Reply:
changed to use this implementation.